File: src/main/java/io/github/justfoxx/teacup/v1/event/EventKey.java
package io.github.justfoxx.teacup.v1.event;

import io.github.justfoxx.teacup.v1.registry.SetRegistry;
import io.github.justfoxx.teacup.v1.utils.ThrowingBiConsumer;
import io.github.justfoxx.teacup.v1.utils.ThrowingConsumer;

import java.util.Set;

public class EventKey <V, I> implements Event<V, I> {
    private final SetRegistry<V> registry;
    private final ThrowingBiConsumer<Set<V>, I> invoker;

    private EventKey(SetRegistry<V> registry, ThrowingBiConsumer<Set<V>, I> invoker) {
        this.registry = registry;
        this.invoker = invoker;
    }

    public static <V,T> EventKey<V,T> of(SetRegistry<V> registry, ThrowingBiConsumer<Set<V>,T> invoker) {
        return new EventKey<>(registry, invoker);
    }

    public static <V> EventKey<V,Void> of(SetRegistry<V> registry, ThrowingConsumer<Set<V>> invoker) {
        return EventKey.of(registry, ThrowingBiConsumer.convert(invoker));
    }

    public V onEvent(V value) {
        registry.add(value);
        return value;
    }

    public Set<V> getAll() {
        return registry.getAll();
    }

    @Override
    public void invoke(I value) {
        invoker.accept(getAll(),value);
    }
}

File: src/main/kotlin/io/github/justfoxx/teacup/v1/event/EventKey.kt
package io.github.justfoxx.teacup.v1.event

import io.github.justfoxx.teacup.v1.registry.SetRegistry


/**
 * An implementation of Event that should be used when registry values don't need additional data.
 *
 * @param V The value type.
 * @param I The invoker additional type (if not present, use Nothing).
 */
class EventKey<V, I> constructor(
    private val registry: SetRegistry<V>,
    private val invoker: (Set<V>, I?) -> Unit
) : Event<V, I> {

    /**
     * Registers a new value in the event registry.
     *
     * @param value The value to register.
     */
    fun onEvent(value: V) {
        registry.add(value)
    }

    /**
     * Returns all the values registered in the event registry.
     * This method shouldn't be used outside the event system.
     *
     * @return All registered values.
     */
    private val all: Set<V>
        get() = registry.all

    /**
     * Invokes the event and passes the given additional value to the invoker.
     * This method shouldn't be used outside the event system.
     *
     * @param value The value to pass to the invoker.
     */
    override operator fun invoke(value: I?) {
        invoker.invoke(all, value)
    }
}